File: Daily/514.freedom-trail.cpp
/*
 * @lc app=leetcode id=514 lang=cpp
 *
 * [514] Freedom Trail
 */
#include <cmath>
#include <string>
#include <vector>

using std::string;
using std::vector;

// @lc code=start
class Solution {
 public:
  int findRotateSteps(const string& ring, const string& key) {
    vector<vector<int>> memo;
    memo.resize(ring.size());
    for (int i = 0; i < memo.size(); i++) {
      memo[i].resize(key.size());
      std::fill(memo[i].begin(), memo[i].end(), -1);
    }

    int result = findRotateSteps(ring, 0, key, 0, memo);
    return result;
  }

 private:
  /*
  Ring is at ring[r], how many steps needed at minimum to fulfill key[k:]
  */
  int findRotateSteps(const string& ring,
                      int r,
                      const string& key,
                      int k,
                      vector<vector<int>>& memo) {
    if (k == key.size()) {
      return 0;  // done
    }

    if (memo[r][k] != -1) {
      return memo[r][k];
    }

    if (ring[r] == key[k]) {
      int steps = 1 + findRotateSteps(ring, r, key, k + 1, memo);
      memo[r][k] = steps;
      return steps;
    }

    int clockwiseR = r;
    int clockwiseSteps = 0;
    while (ring[clockwiseR] != key[k]) {
      clockwiseSteps++;
      clockwiseR--;
      clockwiseR = (clockwiseR + ring.size()) % ring.size();
    }

    int anticlockwiseR = r;
    int anticlockwiseSteps = 0;
    while (ring[anticlockwiseR] != key[k]) {
      anticlockwiseSteps++;
      anticlockwiseR++;
      anticlockwiseR = anticlockwiseR % ring.size();
    }

    int steps =
        std::min(clockwiseSteps +
                     findRotateSteps(ring, clockwiseR, key, k + 1, memo),
                 anticlockwiseSteps +
                     findRotateSteps(ring, anticlockwiseR, key, k + 1, memo)) +
        1;
    memo[r][k] = steps;
    return steps;
  }
};
// @lc code=end

int main() {
  Solution sol;
  string ring = "nyngl";
  string key = "yyynnnnnnlllggg";
  sol.findRotateSteps(ring, key);
}

File: src/2022-job-autumn/514.freedom-trail.js
/*
 * @lc app=leetcode id=514 lang=javascript
 *
 * [514] Freedom Trail
 */

// @lc code=start
/**
 * @param {string} ring
 * @param {string} key
 * @return {number}
 */
var findRotateSteps = function (ring, key) {
  /** @type {Map<string, number[]>} */
  const letterToIndexes = new Map();

  for (let i = 0; i < ring.length; i++) {
    const letter = ring[i];
    const indexes = letterToIndexes.get(letter) ?? [];
    indexes.push(i);
    letterToIndexes.set(ring[i], indexes);
  }

  /** @type {Map<string, number>} */
  const memo = new Map();
  /**
   * 从 ring[ringIndex:] 得到 key[keyIndex:] 最少需要多少步
   * @param {number} ringIndex
   * @param {number} keyIndex
   * @returns {number}
   */
  const dp = (ringIndex, keyIndex) => {
    if (keyIndex === key.length) {
      return 0;
    }
    const memoKey = `${ringIndex}-${keyIndex}`;
    if (memo.has(memoKey)) {
      return memo.get(memoKey);
    }
    const letter = key[keyIndex];

    const indexes = letterToIndexes.get(letter) ?? [];

    let result = Infinity;

    for (const index of indexes) {
      const rotateSteps = Math.min(
        Math.abs(ringIndex - index),
        ring.length - Math.abs(ringIndex - index),
      );
      result = Math.min(result, 1 + rotateSteps + dp(index, keyIndex + 1));
    }

    memo.set(memoKey, result);
    return result;
  };

  const result = dp(0, 0);
  return result;
};
// @lc code=end
